packages/pkl.tar/File.pkl

```diff
@@ -0,0 +1,401 @@
+    }
+    when (!(info.gname is PosixPortableString)) {
+      ["gname"] = info.gname
+    }
+  }
+}
+
+/// The payload bytes of this file, padded to the nearest 512 bytes.
+fixed paddedBytes: Buffer(length % blockLength == 0) = List()
+
+/// Generate the UStar header for this file including [type]-specific customizations.
+function makeHeader(path: String, pax: tar.Pax?): Block = makeBaseHeader(path, pax)
+
+/// Generate the UStar header for this file before [type]-specific customatizations.
+function makeBaseHeader(path: String, pax: tar.Pax?): Block =
+  let (pathParts = splitFilename(path, pax))
+    List(
+      pathParts.first, // name [0, 100)
+      formatOctal(info!!.mode, 8), // mode [100, 108)
+      formatOctal(validateNumericField(info!!.uid, ustarMaxUidGid, "uid", pax), 8), // uid [108, 116)
+      formatOctal(validateNumericField(info!!.gid, ustarMaxUidGid, "gid", pax), 8), // gid [116, 124)
+      nul(12), // size [124, 136)
+      formatOctal(
+        if (info!!.mtime is UInt) validateNumericField(info!!.mtime as UInt, ustarMaxSizeTimestamp, "mtime", pax)
+        else validateNumericField(info!!.mtime.toFloat().toInt(), ustarMaxSizeTimestamp, "mtime", pax),
+        12
+      ), // mtime [136, 148)
+      nul(8), // cksum [148, 156)
+      List(typeByte), // typeflag [156, 157)
+      nul(100), // linkname [157, 257)
+      magic, // magic [257, 263)
+      version, // version [263, 265)
+      validateStringField(info!!.uname, 32, "uname", pax), // uname [265, 297)
+      validateStringField(info!!.gname, 32, "gname", pax), // gname [297, 329)
+      nul(8), // devmajor [329, 337)
+      nul(8), // devminor [337, 345)
+      pathParts.second, // prefix [345, 500)
+      nul(12) // padding [500, 512)
+    ).flatten()
+
+local self = this
+
+/// Generate the pax header and payload for this file, if enabled and applicable.
+function makePax(path: String, pax: tar.Pax): Buffer(length % blockLength == 0) =
+  let (
+    paxHeader = new ExtendedHeader {
+      paxFields = (self.paxFields) {
+        when (splitFilename(path, pax) == null) {
+          ["path"] = path
+        }
+      }
+    }
+  )
+    if (paxHeader.paxFields.isEmpty)
+      List()
+    else
+      checksum(paxHeader.makeHeader(path, pax)) + paxHeader.paddedBytes
+
+/// File metdata information.
+class Info {
+  /// File permissions mode bit set.
+  ///
+  /// Cases:
+  /// * 0o4000 - setuid
+  /// * 0o2000 - setgid
+  /// * 0o1000 - Reserved
+  /// * 0o0400 - Readable by user
+  /// * 0o0200 - Writable by user
+  /// * 0o0100 - Executable by user
+  /// * 0o0040 - Readable by group
+  /// * 0o0020 - Writable by group
+  /// * 0o0010 - Executable by group
+  /// * 0o0004 - Readable by other
+  /// * 0o0002 - Writable by other
+  /// * 0o0001 - Executable by other
+  mode: Int(isBetween(0o0000, 0o7777))
+
+  /// File modification time in epoch seconds.
+  ///
+  /// Truncated to whole seconds for UStar archives.
+  /// Must be < 2**33 for UStar archives.
```

Comment:
This should be updated

Reply:
String is actually better here either way since `Number.toFixed()` doesn't support enough digits to epoch seconds with nanosecond precision.